File: cnvm/tsr.py
import json
import os
import numpy as np

# Lazy-loaded global state (populated by load_slider_map)
_TSR_MAP: dict[str, tuple[int, int]] = {}
_DIM: int = 0


def load_slider_map(manifest_dir: str = "manifest") -> dict[str, tuple[int, int]]:
    """
    Reads sliders.json and builds the flat TSR coordinate map.

    Each slider entry must have:
      - "coordinate": N  (unique integer, 0-based, determines index in state vector)

    Every slider occupies exactly ONE index in the state vector:
      index = coordinate

    This is the flat single-slot design — no block_size, no stride, no padding.
    To add a new concept, add a new entry with the next available coordinate.

    Returns:
        TSR_MAP: dict mapping slider_name -> (index, index) — both values equal
    """
    path = os.path.join(manifest_dir, "sliders.json")
    with open(path, "r") as f:
        data = json.load(f)

    tsr_map: dict[str, tuple[int, int]] = {}
    for coord, (name, entry) in enumerate(data.items()):
        # Each slider is a single slot: start == end == coord
        tsr_map[name] = (coord, coord)

    global _TSR_MAP, _DIM
    _TSR_MAP = tsr_map
    # DIM = total number of addressable registers = length of map
    _DIM = len(tsr_map)

    return tsr_map


def get_tsr_map() -> dict[str, tuple[int, int]]:
    """Returns the currently loaded TSR map. Must call load_slider_map() first."""
    if not _TSR_MAP:
        raise RuntimeError("TSR map not loaded. Call load_slider_map(manifest_dir) first.")
    return _TSR_MAP


def get_dim() -> int:
    """Returns the total hidden dimension (= number of sliders). Must call load_slider_map() first."""
    if _DIM == 0:
        raise RuntimeError("TSR map not loaded. Call load_slider_map(manifest_dir) first.")
    return _DIM


def get_register_offset(name: str) -> int:
    """Returns the index of a named register in the state vector."""
    tsr_map = get_tsr_map()
    if name not in tsr_map:
        raise ValueError(f"Unknown register name: '{name}'. Available: {list(tsr_map.keys())}")
    return tsr_map[name][0]


def get_register_slice(name: str) -> slice:
    """Returns a slice covering this register's index (always a single element)."""
    tsr_map = get_tsr_map()
    if name not in tsr_map:
        raise ValueError(f"Unknown register name: '{name}'")
    idx, _ = tsr_map[name]
    return slice(idx, idx + 1)


def create_empty_state() -> np.ndarray:
    """Returns a zero-initialized state vector of size DIM."""
    return np.zeros(get_dim(), dtype=np.float32)


def set_register_value(state: np.ndarray, name: str, value: float) -> None:
    """Sets the value of a named register in the state vector."""
    idx = get_register_offset(name)
    state[idx] = value


def get_register_value(state: np.ndarray, name: str) -> float:
    """Gets the value of a named register from the state vector."""
    idx = get_register_offset(name)
    return float(state[idx])


def sliders_to_vector(sliders: dict[str, float]) -> np.ndarray:
    """Converts a dict of {slider_name: value} into a DIM-dimensional register vector."""
    state = create_empty_state()
    for name, value in sliders.items():
        if name in get_tsr_map():
            set_register_value(state, name, float(value))
        else:
            raise ValueError(f"Slider '{name}' does not correspond to a known register.")
    return state


def vector_to_sliders(state: np.ndarray) -> dict[str, float]:
    """Extracts all named register values from a state vector."""
    tsr_map = get_tsr_map()
    return {
        name: get_register_value(state, name)
        for name in tsr_map
        if name not in ("META::PROVENANCE", "META::EVIDENCE")
    }


def format_state(state: np.ndarray) -> str:
    """Returns a clean string summary of active registers for logging/debugging."""
    tsr_map = get_tsr_map()
    lines = ["TSR Hidden State:"]
    for name, (idx, _) in tsr_map.items():
        val = state[idx]
        if abs(val) > 1e-5:
            lines.append(f"  {name:<35} [{idx:02d}]: {val:.4f}")
    return "\n".join(lines)

File: dashboard/src/cnvm/tsr.ts
export interface SlidersConfigEntry {
  name: string;
  description: string;
  // block_size is intentionally removed — every slider is exactly one index
}

export type SlidersConfig = Record<string, SlidersConfigEntry>;

export interface TsrMapInfo {
  tsrMap: Record<string, [number, number]>;
  dim: number;
}

/**
 * Builds the flat TSR coordinate map from sliders.json.
 *
 * Flat single-slot design: every slider occupies exactly one index.
 *   index = order in slidersConfig
 *   dim   = total number of entries
 *
 * No block_size, no stride, no padding. Adding a new concept = adding one entry.
 */
export function loadSliderMap(slidersConfig: SlidersConfig): TsrMapInfo {
  const tsrMap: Record<string, [number, number]> = {};
  let coord = 0;

  for (const [name, entry] of Object.entries(slidersConfig)) {
    // Single slot: start == end == coordinate
    tsrMap[name] = [coord, coord];
    coord++;
  }

  const dim = coord;
  return { tsrMap, dim };
}

export function createEmptyState(dim: number): number[] {
  return new Array(dim).fill(0);
}

export function setRegisterValue(
  state: number[],
  tsrMap: Record<string, [number, number]>,
  name: string,
  value: number
): void {
  if (!(name in tsrMap)) {
    throw new Error(`Unknown register name: ${name}`);
  }
  const [idx] = tsrMap[name];
  state[idx] = value;
}

export function getRegisterValue(
  state: number[],
  tsrMap: Record<string, [number, number]>,
  name: string
): number {
  if (!(name in tsrMap)) {
    throw new Error(`Unknown register name: ${name}`);
  }
  const [idx] = tsrMap[name];
  return state[idx];
}

export function slidersToVector(
  sliders: Record<string, number>,
  tsrMap: Record<string, [number, number]>,
  dim: number
): number[] {
  const state = createEmptyState(dim);
  for (const [name, val] of Object.entries(sliders)) {
    if (name in tsrMap) {
      setRegisterValue(state, tsrMap, name, val);
    }
  }
  return state;
}

export function vectorToSliders(
  state: number[],
  tsrMap: Record<string, [number, number]>
): Record<string, number> {
  const sliders: Record<string, number> = {};
  const metaNames = new Set(["META::PROVENANCE", "META::EVIDENCE", "RESERVED"]);
  for (const name of Object.keys(tsrMap)) {
    if (!metaNames.has(name)) {
      sliders[name] = getRegisterValue(state, tsrMap, name);
    }
  }
  return sliders;
}

/**
 * Returns the set of register indices that are 'semantic' (included in
 * layer normalisation and clipping). Excludes META::PROVENANCE and META::EVIDENCE.
 */
export function getSemanticIndices(
  tsrMap: Record<string, [number, number]>,
  dim: number
): number[] {
  const metaNames = new Set(["META::PROVENANCE", "META::EVIDENCE", "RESERVED"]);
  const metaIndices = new Set<number>();

  for (const [name, [start, end]] of Object.entries(tsrMap)) {
    if (metaNames.has(name)) {
      for (let i = start; i <= end; i++) {
        metaIndices.add(i);
      }
    }
  }

  const result: number[] = [];
  for (let i = 0; i < dim; i++) {
    if (!metaIndices.has(i)) {
      result.push(i);
    }
  }
  return result;
}
